### PR TITLE
Report error when casting from tuple of ambiguous type.

### PR DESCRIPTION
**Pull request by mbaulch**
Fixes https://github.com/nim-lang/Nim/issues/4346.

### CODE REVIEW

File: compiler/semstmts.nim [hunk 1]
```diff
@@ -455,6 +455,13 @@ proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
         if hasEmpty(typ):
           localError(def.info, errCannotInferTypeOfTheLiteral,
                      ($typ.kind).substr(2).toLower)
+        elif typ.kind == tyTuple and def.kind == nkConv:
```
Comment (Araq) [inline compiler/semstmts.nim]: wrong position, it has nothing to do with the 'var' context, it's the type conversion missing these checks.

**Comment by mbaulch** [comment]
You're right. All conversions need to be checked. I don't know what I was thinking! Checking in the right place, the solution is much more elegant. Thanks.

**Comment by Araq** [comment]
Why is it illegal though? You can also annotate the expression with the correct types instead.

**Comment by mbaulch** [comment]
It doesn't have to be illegal. You suggested in https://github.com/nim-lang/Nim/issues/3516#issuecomment-166273578 that
```nim
var
  foo: seq[int]
  bar: seq[int]

(foo, bar) = (@[], @[])
```

can be rejected, so I figured such a tuple conversion should fail too. I think it'd help to have some clear guidelines about what forms of type inference are supported. This seems to be a grey area.

Anyway, I've implemented this as you suggested. See https://github.com/nim-lang/Nim/pull/4488.

**Pull request closed without merging**